### ISSUE TITLE
discrepancy between 1 core and 8 core computation results.

### ISSUE THREAD

**Issue by taylanbil** [pytorch/xla#2441]
## 🐛 Bug

Slicing the input tensor changes linear layer output significantly.

## To Reproduce

1. Put these in `smallrepro.py`:
```python
import torch
import numpy as np
import torch.nn as nn
import torch_xla.core.xla_model as xm
import torch_xla.distributed.xla_multiprocessing as xmp


def _summary(prefix, t):
    dat = t.detach()
    dat = [dat.sum(), dat.shape, xm.get_ordinal()]
    out = '{} - ' + "{} "*len(dat)
    return out.format(prefix, *dat)

def summarize(prefix, t, master_only=False):
    out = []
    if isinstance(t, list):
        out.extend(_summary(prefix, _) for _ in t)
    else:
        out.append(_summary(prefix, t))
    xm.rendezvous('hi')
    if master_only:
        xm.master_print('\n'.join(out))
    else:
        print('\n'.join(out))
    xm.rendezvous('hi')

def main(*a):
    n, m = 13, 2
    gbsz, wsz = 8, xm.xrt_world_size()
    mbsz = gbsz // wsz
    device = xm.xla_device()
    np.random.seed(1000)
    torch.manual_seed(1000)
    #i = torch.ones(8, n).to(device)
    i = torch.randn(gbsz, n).to(device)
    if wsz == 8:
        i = torch.narrow(i, 0, xm.get_ordinal()*mbsz, mbsz)
    LL = LL.to(device)
    out = LL(i)
    summarize('LAYERW', LL.weight, master_only=True)
    summarize('LAYERB', LL.bias, master_only=True)
    summarize('IN', [_ for _ in i])
    summarize('OUT', [_ for _ in out])
    xm.rendezvous('hi')


if __name__ == '__main__':
    import sys
    n = sys.argv[1]
    xmp.spawn(main, args=(1,), nprocs=int(n))
```
2. Put these in `smallrepro.sh`:
```
export XRT_TPU_CONFIG="tpu_worker;0;$TPU_IP_ADDRESS:8470"

for ncores in 1 8
do
    echo ''
    echo "Running $ncores core"
    echo ''
    python smallrepro.py $ncores | sort -n
done
```
3. Output:
```
Running 1 core

IN - 0.9905535578727722 torch.Size([13]) 0
IN - -2.0730087757110596 torch.Size([13]) 0
IN - -2.4356441497802734 torch.Size([13]) 0
IN - 2.6724281311035156 torch.Size([13]) 0
IN - 3.6933655738830566 torch.Size([13]) 0
IN - 3.844777822494507 torch.Size([13]) 0
IN - -4.635951519012451 torch.Size([13]) 0
IN - -7.289895534515381 torch.Size([13]) 0
LAYERB - 0.046467870473861694 torch.Size([2]) 0
LAYERW - -1.977316975593567 torch.Size([2, 13]) 0
OUT - -0.13582393527030945 torch.Size([2]) 0
OUT - 0.16633757948875427 torch.Size([2]) 0
OUT - 0.3280348479747772 torch.Size([2]) 0
OUT - 0.6423056125640869 torch.Size([2]) 0
OUT - -0.7382650375366211 torch.Size([2]) 0
OUT - -0.8302066326141357 torch.Size([2]) 0
OUT - 1.1669577360153198 torch.Size([2]) 0
OUT - 1.2597312927246094 torch.Size([2]) 0

Running 8 core

IN - 0.9905535578727722 torch.Size([13]) 0
IN - -2.0730087757110596 torch.Size([13]) 3
IN - -2.4356441497802734 torch.Size([13]) 7
IN - 2.6724281311035156 torch.Size([13]) 6
IN - 3.6933655738830566 torch.Size([13]) 1
IN - 3.844777822494507 torch.Size([13]) 4
IN - -4.635951519012451 torch.Size([13]) 2
IN - -7.289895534515381 torch.Size([13]) 5
LAYERB - 0.046467870473861694 torch.Size([2]) 0
LAYERW - -1.977316975593567 torch.Size([2, 13]) 0
OUT - -0.13569559156894684 torch.Size([2]) 3
OUT - 0.16423645615577698 torch.Size([2]) 0
OUT - 0.3262776732444763 torch.Size([2]) 6
OUT - 0.6420916318893433 torch.Size([2]) 5
OUT - -0.7397018671035767 torch.Size([2]) 1
OUT - -0.8290789723396301 torch.Size([2]) 4
OUT - 1.166985034942627 torch.Size([2]) 2
OUT - 1.2597153186798096 torch.Size([2]) 7
```
Analyzing this output w/ my custom script that puts them side by side and prints deltas, we see inputs and weights are the same, but outputs are different. 
```
IN      0.9905535578727722      0.9905535578727722      0.00 PCT
IN      -2.0730087757110596     -2.0730087757110596     -0.00 PCT
IN      -2.4356441497802734     -2.4356441497802734     -0.00 PCT
IN      2.6724281311035156      2.6724281311035156      0.00 PCT
IN      3.6933655738830566      3.6933655738830566      0.00 PCT
IN      3.844777822494507       3.844777822494507       0.00 PCT
IN      -4.635951519012451      -4.635951519012451      -0.00 PCT
IN      -7.289895534515381      -7.289895534515381      -0.00 PCT
LAYERB  0.046467870473861694    0.046467870473861694    0.00 PCT
LAYERW  -1.977316975593567      -1.977316975593567      -0.00 PCT
OUT     -0.13582393527030945    -0.13569559156894684    0.09 PCT
OUT     0.16633757948875427     0.16423645615577698     1.26 PCT
OUT     0.3280348479747772      0.3262776732444763      0.54 PCT
OUT     0.6423056125640869      0.6420916318893433      0.03 PCT
OUT     -0.7382650375366211     -0.7397018671035767     -0.19 PCT
OUT     -0.8302066326141357     -0.8290789723396301     0.14 PCT
OUT     1.1669577360153198      1.166985034942627       -0.00 PCT
OUT     1.2597312927246094      1.2597153186798096      0.00 PCT
```

## Expected behavior

The python script 1 core and 8 core computations should be the same.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: TPU
 - torch_xla version: nightly, 1.6

## Additional context

I believe this is the reason behind DLRM not converging as well as it should. Training DLRM w/ 1 core converges the same way as GPUs, whereas training on 8 cores does not converge as well, loss value starts around 0.52s and goes down to 0.45's, not 0.44's.

Analyzing this lack of convergence, I experimented with disabling the backward passes and I observed that the loss values output are different in 1 core vs 8 cores. Then I looked at intermediate tensors, and observed that bottom and top mlps produce weirdly different results, and the delta is amplified with each layer, and the outputs are significantly different. So even though the differences in this repro is small, it gets amplified w/ depth in a real application.

**Comment by taylanbil** [comment]
Even removing the slice, I'm getting slightly different results:

```python
import torch
import numpy as np
import torch.nn as nn
import torch_xla.core.xla_model as xm
import torch_xla.distributed.xla_multiprocessing as xmp


def _summary(prefix, t):
    dat = t.detach()
    dat = [dat.sum(), dat.shape, xm.get_ordinal()]
    out = '{} - ' + "{} "*len(dat)
    return out.format(prefix, *dat)

def summarize(prefix, t, master_only=False):
    out = []
    if isinstance(t, list):
        out.extend(_summary(prefix, _) for _ in t)
    else:
        out.append(_summary(prefix, t))
    xm.rendezvous('hi')
    if master_only:
        xm.master_print('\n'.join(out))
    else:
        print('\n'.join(out))
    xm.rendezvous('hi')

def main(*a):
    n, m = 13, 2
    #n, m = 64, 353
    gbsz, wsz = 8, xm.xrt_world_size()
    mbsz = gbsz // wsz
    device = xm.xla_device()
    np.random.seed(100)
    torch.manual_seed(100)
    LL = nn.Linear(n, m, bias=True)
    i = torch.ones(mbsz, n).to(device)
    LL = LL.to(device)
    out = LL(i)
    summarize('LAYERW', LL.weight, master_only=True)
    summarize('LAYERB', LL.bias, master_only=True)
    summarize('IN', [_ for _ in i])
    summarize('OUT', [_ for _ in out])
    xm.rendezvous('hi')


if __name__ == '__main__':
    import sys
    n = sys.argv[1]
    xmp.spawn(main, args=(1,), nprocs=int(n))
```

output:
```
Running 1 core

IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 0
LAYERB - 0.12962612509727478 torch.Size([2]) 0
LAYERW - 0.3432807922363281 torch.Size([2, 13]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0
OUT - 0.46916472911834717 torch.Size([2]) 0

Running 8 core

IN - 13.0 torch.Size([13]) 0
IN - 13.0 torch.Size([13]) 1
IN - 13.0 torch.Size([13]) 2
IN - 13.0 torch.Size([13]) 3
IN - 13.0 torch.Size([13]) 4
IN - 13.0 torch.Size([13]) 5
IN - 13.0 torch.Size([13]) 6
IN - 13.0 torch.Size([13]) 7
LAYERB - 0.12962612509727478 torch.Size([2]) 0
LAYERW - 0.3432807922363281 torch.Size([2, 13]) 0
OUT - 0.4729069471359253 torch.Size([2]) 0
OUT - 0.4729069471359253 torch.Size([2]) 1
OUT - 0.4729069471359253 torch.Size([2]) 2
OUT - 0.4729069471359253 torch.Size([2]) 3
OUT - 0.4729069471359253 torch.Size([2]) 4
OUT - 0.4729069471359253 torch.Size([2]) 5
OUT - 0.4729069471359253 torch.Size([2]) 6
OUT - 0.4729069471359253 torch.Size([2]) 7
```

About 1% different here.

**Comment by taylanbil** [comment]
I'm trying to repro the full case I'm seeing w/ dlrm, where every linear layer amplifies the delta between the outputs, but no luck so far.

**Comment by taylanbil** [comment]
DLRM convergence issue was related to another bug, turns out this issue is not related to that. At this point it's not clear if the small delta b/w 1 core and 8 core computations lead to an actual real life problem. Closing.
